### PR TITLE
Ads v3

### PR DESCRIPTION
**Pull request by EItanya**
# Description

Continue migration to Envoy V3 api.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make update-deps generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo [contribution guide](https://docs.solo.io/gloo/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: install/helm/gloo/templates/8-gateway-proxy-service-account.yaml [hunk 1]
```diff
@@ -11,4 +11,6 @@ metadata:
 automountServiceAccountToken: false
 {{ end }}
 
-{{ end }}
\ No newline at end of file
+{{ end }}
+---
+
```
Comment (kdorosh) [inline install/helm/gloo/templates/8-gateway-proxy-service-account.yaml:16]: remove this?

File: projects/gloo/cli/pkg/xdsinspection/get_last_config.go [hunk 2]
```diff
@@ -292,9 +293,11 @@ func (xd *XdsDump) String() string {
...
 func toYaml(pb proto.Message) ([]byte, error) {
-	jsn, err := protoutils.MarshalBytes(pb)
+	buf := &bytes.Buffer{}
+	jpb := &jsonpb.Marshaler{}
+	err := jpb.Marshal(buf, pb)
 	if err != nil {
 		return nil, err
 	}
-	return yaml.JSONToYAML(jsn)
+	return yaml.JSONToYAML(buf.Bytes())
```
Comment (kdorosh) [inline projects/gloo/cli/pkg/xdsinspection/get_last_config.go:302]: why this change?

Comment (EItanya) [reply projects/gloo/cli/pkg/xdsinspection/get_last_config.go:302]: It was using the gogo marshaler, which wasn't working, they are go protos

File: projects/gloo/pkg/syncer/envoy_translator_syncer.go [hunk 3]
```diff
@@ -48,15 +49,13 @@ func init() {
...
-		emptyResource,
-		emptyResource,
-		emptyResource,
+	emptySnapshot = cache_v3.NewSnapshot(
+		emptyVersionKey,
+		nil,
+		nil,
+		nil,
+		nil,
+		nil,
```
Comment (kdorosh) [inline projects/gloo/pkg/syncer/envoy_translator_syncer.go:58]: this doesn't look right at first glance. why 5 nils? if it's for each resource type, we should comment each nil

File: projects/gloo/pkg/syncer/translator_syncer_test.go [hunk 4]
```diff
@@ -109,42 +108,41 @@ var _ = Describe("Translate Proxy", func() {
...
-		Expect(xdsCache.setSnap).To(BeEquivalentTo(sanitizer.snap))
-	})
+	// It("updates the cache with the sanitized snapshot", func() {
+	// 	sanitizer.snap = envoycache.NewEasyGenericSnapshot("easy")
+	// 	err := syncer.Sync(context.Background(), snap)
+	// 	Expect(err).NotTo(HaveOccurred())
+	//
+	// 	Expect(sanitizer.called).To(BeTrue())
+	// 	Expect(xdsCache.setSnap).To(BeEquivalentTo(sanitizer.snap))
+	// })
```
Comment (kdorosh) [inline projects/gloo/pkg/syncer/translator_syncer_test.go:118]: we should fix and put back

File: projects/gloo/pkg/plugins/upstreamconn/plugin.go [hunk 5]
```diff
@@ -14,6 +14,10 @@ import (
 	"github.com/solo-io/gloo/projects/gloo/pkg/plugins"
 )
 
+var (
+	_ plugins.UpstreamPlugin = new(Plugin)
```
Comment (mlholland) [inline projects/gloo/pkg/plugins/upstreamconn/plugin.go:18]: I don't see any uses of UpstreamPlugin. Why is it needed?

Comment (kdorosh) [reply projects/gloo/pkg/plugins/upstreamconn/plugin.go:18]: this creates a compile-time assertion that `Plugin` implements the the `UpstreamPlugin` interface

Comment (kdorosh) [reply projects/gloo/pkg/plugins/upstreamconn/plugin.go:18]: in the plugin translation loop we take each registered plugin, and if it properly implements the interface, we run the functions defined on the plugin. this is a safer check to make sure we run the plugin

Comment (EItanya) [reply projects/gloo/pkg/plugins/upstreamconn/plugin.go:18]: Great question, this is essentially the accepted way of doing `go` compile time assertions on interfaces that aren't explicitly returned. For instance. the `Plugin` in this file is an `plugins.UpstreamPlugin`, but it is not returned as such, and therefore there is no compile time assurance that it is in fact a `plugins.UpstreamPlugin`. These plugins are added as just `plugins.Plugin` in the setup, but then called in a generic way [here](https://github.com/solo-io/gloo/blob/004febcfbd442f6f369b2ce14f89a690f4fabc65/projects/gloo/pkg/translator/clusters.go#L49). Therefore, without this line there would be no way during compile time to check whether this method would get called.

File: pkg/utils/gogoutils/health_check.go [hunk 6]
```diff
@@ -149,15 +148,15 @@ func ToEnvoyHealthCheck(check *envoycore_gloo.HealthCheck) (*envoycore.HealthChe
...
 			if err != nil {
 				return nil, err
 			}
 			hc.HealthChecker = &envoycore.HealthCheck_CustomHealthCheck_{
 				CustomHealthCheck: &envoycore.HealthCheck_CustomHealthCheck{
 					Name: typed.CustomHealthCheck.GetName(),
-					ConfigType: &envoycore.HealthCheck_CustomHealthCheck_TypedConfig{
-						TypedConfig: converted,
+					ConfigType: &envoycore.HealthCheck_CustomHealthCheck_HiddenEnvoyDeprecatedConfig{
+						HiddenEnvoyDeprecatedConfig: converted,
```
Comment (arianaw66) [inline pkg/utils/gogoutils/health_check.go:159]: Does this break if we make it a TypedConfig? Seems like a conversion we'll need eventually.

File: projects/gloo/cli/pkg/xdsinspection/get_last_config.go [hunk 7]
```diff
@@ -137,10 +142,6 @@ func getXdsDump(ctx context.Context, xdsPort, proxyName, proxyNamespace string)
 				if filter.Name == "envoy.http_connection_manager" {
 					var hcm envoyhttp.HttpConnectionManager
 					switch config := filter.ConfigType.(type) {
-					case *envoylistener.Filter_Config:
-						if err := envoyutil.StructToMessage(config.Config, &hcm); err == nil {
-							hcms = append(hcms, hcm)
-						}
```
Comment (arianaw66) [inline projects/gloo/cli/pkg/xdsinspection/get_last_config.go:143]: Should we still include the HiddenEnvoyDeprecatedConfig case, which in theory can still be hit?

File: pkg/utils/gogoutils/health_check.go (hunk 6 above)
Comment (EItanya) [reply pkg/utils/gogoutils/health_check.go:159]: @yuval-k would this break envoy? In order to remove this we would have to update our API which would technically be a break, or we could potentially just convert our structs to anys

File: projects/gloo/cli/pkg/xdsinspection/get_last_config.go (hunk 7 above)
Comment (EItanya) [reply projects/gloo/cli/pkg/xdsinspection/get_last_config.go:143]: Can it? I don't think we're setting it anywhere

File: projects/gloo/pkg/plugins/upstreamconn/plugin.go (hunk 5 above)
Comment (mlholland) [reply projects/gloo/pkg/plugins/upstreamconn/plugin.go:18]: An excellent explanation. Thanks!

Comment (kdorosh) [reply projects/gloo/pkg/plugins/upstreamconn/plugin.go:18]: @GrahamGoudeau makes a good point that we could improve this by casting a null to the type rather than calling `new` so we don't have to instantiate a new object here but still get the assertion

File: projects/gloo/cli/pkg/xdsinspection/get_last_config.go (hunk 7 above)
Comment (arianaw66) [reply projects/gloo/cli/pkg/xdsinspection/get_last_config.go:143]: I'm wondering that too, but it seems better to be thorough in a case statement.

File: pkg/utils/gogoutils/health_check.go (hunk 6 above)
Comment (arianaw66) [reply pkg/utils/gogoutils/health_check.go:159]: I didn't meant to change our API. I meant to take what the user provides us as a Config and give it to Envoy as a TypedConfig. Same thing we do with Regex to SafeRegex.

File: projects/gloo/pkg/plugins/upstreamconn/plugin.go (hunk 5 above)
Comment (EItanya) [reply projects/gloo/pkg/plugins/upstreamconn/plugin.go:18]: We could also create private interfaces combining all of the extensions and return that

**Comment by arianaw66** [comment]
Can we add this to the linkerd plugin? In my first PR, this plugin wasn't implementing RoutePlugin, and it wasn't caught until the kube2e tests. https://github.com/solo-io/gloo/pull/3110/commits/dd2b8b50a8d2c521be23c9deb6d6e2e654a690ed
We can also add this separately if it's easier.
```
var (
	_ plugins.RoutePlugin = new(Plugin)
	_ plugins.UpstreamPlugin = new(Plugin)
)
```

File: test/e2e/ratelimit_test.go [hunk 8]
```diff
@@ -48,7 +48,7 @@ func (this *acceptOrDenyRateLimitServer) ShouldRateLimit(ctx context.Context, re
 	}
 }
 
-var _ = Describe("Rate Limit", func() {
+var _ = FDescribe("Rate Limit", func() {
```
Comment (arianaw66) [inline test/e2e/ratelimit_test.go:51]:
```suggestion
var _ = Describe("Rate Limit", func() {
```

**Comment by arianaw66** [comment]
🤞

File: test/e2e/fault_injection_test.go [hunk 9]
```diff
@@ -48,19 +49,19 @@ var _ = Describe("Fault Injection", func() {
 			return err
 		}
 
-		envoyPort := services.NextBindPort()
+		// envoyPort := services.NextBindPort()
```
Comment (kdorosh) [inline test/e2e/fault_injection_test.go:52]: remove this?

File: test/e2e/fault_injection_test.go [hunk 10]
```diff
@@ -48,19 +49,19 @@ var _ = Describe("Fault Injection", func() {
...
 			Expect(err).NotTo(HaveOccurred())
 			Eventually(func() error {
-				_, err := http.Get(fmt.Sprintf("http://%s:%d/status/200", "localhost", envoyPort))
+				_, err := http.Get(fmt.Sprintf("http://%s:%d/status/200", "localhost", defaults.HttpPort))
 				if err != nil {
 					return err
 				}
 				return nil
-			}, "20s", ".1s").Should(BeNil())
+			}, "5m", ".1s").Should(BeNil())
```
Comment (kdorosh) [inline test/e2e/fault_injection_test.go:64]: we can probably increase the polling interval here if we're upping the timeout to 5m

**Comment by kdorosh** [comment]
/kick after fixing regression test cluster api server

File: docs/content/guides/dev/plugin.go [hunk 11]
```diff
@@ -16,8 +16,6 @@ import (
 	"github.com/solo-io/solo-kit/pkg/api/v1/clients"
 	"github.com/solo-io/solo-kit/pkg/api/v1/resources/core"
 
-	// add these imports to use Envoy's API
-	envoyapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
 	envoycore "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
```
Comment (kdorosh) [inline docs/content/guides/dev/plugin.go:19]: these aren't v3, and the changes to this file are docs-related (so we should double check to make sure this is up to date with our real plugins)

**Comment by kdorosh** [comment]
/kick regression test flake, aware of and have seen before https://storage.googleapis.com/solo-public-build-logs/logs.html?buildid=1f003b20-ab30-4344-831b-cbaa741daee9

File: pkg/utils/gogoutils/health_check.go (hunk 6 above)
Comment (yuval-k) [reply pkg/utils/gogoutils/health_check.go:159]: it might break, not sure - why are we using HiddenEnvoyDeprecatedConfig?

**Comment by arianaw66** [comment]
Is this PR still in-flight? Maybe we should close or label WIP? @EItanya @kdorosh

**Comment by kdorosh** [comment]
Happy to leave open until we support v2 and v3 in gloo simultaneously. It may prove useful once we are ready to start that work

**Comment by kdorosh** [comment]
closing as stale, v3 API work has been merged https://github.com/solo-io/gloo/pull/3894
